### PR TITLE
Add floating point variable types: Float16, Float8, BFloat16, TensorFloat32, Float6, Float4 (ESROGUE-730)

### PR DESCRIPTION
**Pull request by ruck314**
## Summary

Extends Rogue's variable model system with six new floating point types for NVIDIA GPU interoperability.

### New types

| Type | Format | Bits | Value Range | NVIDIA Support |
|------|--------|------|-------------|----------------|
| `pr.Float16` / `pr.Float16BE` | IEEE 754 half-precision, 1s/5e/10m | 16 | ±65504.0 | All |
| `pr.Float8` / `pr.Float8BE` | E4M3, NaN=0x7F, no Inf | 8 | ±448.0 | Hopper, Blackwell |
| `pr.BFloat16` / `pr.BFloat16BE` | 1s/8e/7m, IEEE exponent range | 16 | ±3.39e38 | Ampere, Hopper, Blackwell |
| `pr.TensorFloat32` / `pr.TensorFloat32BE` | 1s/8e/10m, 19 bits in 4 bytes | 19 (32 storage) | ±3.40e38 | Ampere, Hopper, Blackwell |
| `pr.Float6` / `pr.Float6BE` | E3M2, no NaN/Inf | 6 | ±28.0 | Blackwell |
| `pr.Float4` / `pr.Float4BE` | E2M1, no NaN/Inf | 4 | ±6.0 | Blackwell |

### Implementation

Each type follows the same full-stack pattern:
- C++ model ID constant in `Constants.h`
- Block get/set methods with inline bit-manipulation converters in `Block.cpp`
- Variable dispatch in `Variable.cpp`
- Python `Model` class with `toBytes`/`fromBytes`/`fromString`/`minValue`/`maxValue`
- Sphinx API documentation (per-type page + consolidated summary reference)

Special-value handling per spec:
- **Float16**: full IEEE 754 NaN and ±Inf
- **Float8**: NaN = `0x7F`, no infinity (clamps to max finite)
- **BFloat16/TF32**: full IEEE 754 NaN and ±Inf
- **Float6/Float4**: no NaN or Inf (clamps to max finite on encode)

### Documentation

Added `docs/src/pyrogue_tree/core/float_types_summary.rst` — a consolidated quick-reference page with format details table, special-value handling, NVIDIA architecture support matrix, and usage example.

## Test plan

- [x] `pytest tests/core/test_float16.py` — all boundary values, round-trip, metadata, remote variable integration
- [x] `pytest tests/core/test_float8.py` — all boundary values, round-trip, metadata, remote variable integration
- [x] `pytest tests/core/test_bfloat16.py` — all boundary values, round-trip, metadata, remote variable integration
- [x] `pytest tests/core/test_tensorfloat32.py` — all boundary values, round-trip, metadata, remote variable integration
- [x] `pytest tests/core/test_float6.py` — all 64 bit patterns, boundary values, NaN/Inf clamping, remote variable integration
- [x] `pytest tests/core/test_float4.py` — all 16 bit patterns, boundary values, NaN/Inf clamping, remote variable integration
- [x] Full regression: 207 passed
- [x] flake8 + cpplint clean

Resolves ESROGUE-730.

### CODE REVIEW

File: src/rogue/interfaces/memory/Variable.cpp [hunk 1]
```diff
@@ -278,6 +278,10 @@ rim::Variable::Variable(std::string name,
             setFloat_ = &rim::Block::setFloat;
             break;
 
+        case rim::Float16:
+            setFloat16_ = &rim::Block::setFloat16;
+            break;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Variable.cpp:295]: Valid. Added `setFloat16_ = NULL;` and `getFloat16_ = NULL;` to the initialization block alongside all other function pointers. Also found and fixed the same omission for `setFloat8_`/`getFloat8_` which was added in this same PR. Fixed in commit 19691ae13.

File: src/rogue/interfaces/memory/Block.cpp [hunk 2]
```diff
@@ -1723,6 +1723,216 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    uint16_t sign     = (f >> 16) & 0x8000;
+    int32_t  exponent = ((f >> 23) & 0xFF) - 127 + 15;
+    uint32_t mantissa = f & 0x7FFFFF;
+
+    if (exponent <= 0) {
+        if (exponent < -10) return sign;
+        mantissa |= 0x800000;
+        uint32_t shift = 14 - exponent;
+        mantissa >>= shift;
+        return sign | (mantissa >> 13);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid. After `mantissa >>= shift`, the variable already holds the 10-bit half-precision mantissa — the subsequent `>> 13` was wrong and flushed all subnormals to zero. Removed it. Kept plain truncation (no round-to-nearest) which is consistent with the normal-path encoding. Fixed in commit 19691ae13.

File: src/rogue/interfaces/memory/Block.cpp [hunk 3]
```diff
@@ -1723,6 +1723,216 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    } else if (exponent == 0xFF - 127 + 15) {
+        if (mantissa) return sign | 0x7C00 | (mantissa >> 13);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid. When a float32 NaN has its payload entirely in the lower 13 bits (e.g. `0x7F800001`), `mantissa >> 13 == 0` and the result was `0x7C00` (infinity). Fixed by forcing the half mantissa LSB to 1 whenever the shifted payload is zero. Fixed in commit 19691ae13.

File: src/rogue/interfaces/memory/Block.cpp [hunk 4]
```diff
@@ -1723,6 +1723,216 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+        if (mantissa == 0) {
+            f = sign;
+        } else {
+            exponent = 1;
+            while (!(mantissa & 0x400)) {
+                mantissa <<= 1;
+                exponent--;
+            }
+            mantissa &= 0x3FF;
+            f = sign | ((exponent + 127 - 15) << 23) | (mantissa << 13);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid. Decrementing a `uint32_t` starting at 1 wraps to `UINT32_MAX` after the second iteration, which corrupts the rebiased exponent for any half-precision subnormal requiring more than one normalization shift. Replaced with an `int32_t` working variable (`exponentWork`) as suggested. Fixed in commit 19691ae13.

File: tests/core/test_model_variables.py [hunk 5]
```diff
@@ -157,3 +168,23 @@ def test_remote_variable_metadata_exposes_model_constraints():
...
+        assert math.isclose(root.Dev.Float16Var.get(), 65504.0, rel_tol=0.0, abs_tol=1.0)
+
+        # Negative max
+        root.Dev.Float16Var.set(-65504.0)
+        assert math.isclose(root.Dev.Float16Var.get(), -65504.0, rel_tol=0.0, abs_tol=1.0)
+
+        # Small value (tests subnormal handling through the C++ path)
+        root.Dev.Float16Var.set(0.0001)
+        result = root.Dev.Float16Var.get()
+        assert math.isclose(result, 0.0001, rel_tol=0.1)
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_model_variables.py]: Valid. Added a test for the smallest positive half-precision subnormal (`2**-24`) which exercises the `floatToHalf`/`halfToFloat` denormal path through the C++ Block code. This test would have caught the subnormal bugs fixed in commit 19691ae13. Fixed in the same commit.

**Comment by ruck314** [comment]
Follow-up to the Copilot review fixes (commit 19691ae13): audited all other new floating point types for the same 5 bug categories.

**Bug 4 (uint32_t underflow in normalization loop) also affected `float8ToFloat`** — and had a related secondary bug: the loop terminated when bit 2 was set (`0x04`) instead of bit 3 (`0x08`), stopping one shift early. The combination caused every Float8 subnormal to decode at 2× the correct value (e.g. `0x01` → `2^-8` instead of `2^-9`). Fixed in commit d42516fda.

**Float6 and Float4 are immune** to both issues — their `ToFloat` decoders use direct arithmetic instead of a normalization loop.

**BFloat16 and TensorFloat32** have an analogous NaN-payload-truncation edge case (Bug 3): a float32 NaN with payload bits only in the lower 16/13 bits will round-trip as Inf rather than NaN. This is an inherent property of the truncation-based encoding (both formats are defined as bit-truncations of float32), not a coding error.

**Bug 1 (NULL init)** for Float8 pointers was already caught and fixed in 19691ae13 alongside the Float16 fix.

File: python/pyrogue/_Model.py [hunk 6]
```diff
@@ -781,6 +781,524 @@ class DoubleBE(Double):
...
+        bits = struct.unpack('I', struct.pack('f', v))[0]
+        bf16 = bits >> 16
+        return bytearray(struct.pack('<H', bf16))
+
+    def fromBytes(self, ba: bytes) -> float:
+        """Decode 2-byte BFloat16 encoding to float.
+
+        Reconstructs float32 by shifting the 16-bit pattern left by 16.
+        """
+        bf16 = struct.unpack('<H', ba[:2])[0]
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py]: Valid — fixed in 67b498132. Both  and  now use  to select  or  when packing/unpacking the uint16, so  correctly produces big-endian bytes.

File: python/pyrogue/_Model.py [hunk 7]
```diff
@@ -781,6 +781,524 @@ class DoubleBE(Double):
...
+    def toBytes(self, value: float) -> bytearray:
+        """Convert float to 2-byte BFloat16 encoding.
+
+        BFloat16 is the upper 16 bits of the float32 bit pattern.
+        All special values (NaN, infinity, zero, subnormals) are preserved.
+        """
+        v = float(value)
+        # Get float32 bit pattern as uint32, take upper 16 bits
+        bits = struct.unpack('I', struct.pack('f', v))[0]
+        bf16 = bits >> 16
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:1006]: Valid — fixed in 67b498132. After computing ,  now checks for  and forces  to keep the NaN payload non-zero.

File: python/pyrogue/_Model.py [hunk 8]
```diff
@@ -781,6 +781,524 @@ class DoubleBE(Double):
...
+
+    def fromBytes(self, ba: bytes) -> float:
+        """Decode 4-byte TensorFloat32 encoding to float.
+
+        TF32 bit pattern is a valid float32 with lower 13 mantissa bits
+        zeroed. Direct reinterpretation as float32 is sufficient.
+        """
+        tf32 = struct.unpack('<I', ba[:4])[0]
+        return struct.unpack('f', struct.pack('I', tf32))[0]
+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:1095]: Valid — fixed in 67b498132. Both  and  now use  to select  or  when packing/unpacking the uint32, so  correctly produces big-endian bytes.

File: python/pyrogue/_Model.py [hunk 9]
```diff
@@ -781,6 +781,524 @@ class DoubleBE(Double):
...
+    def toBytes(self, value: float) -> bytearray:
+        """Convert float to 4-byte TensorFloat32 encoding.
+
+        TF32 zeros the lower 13 mantissa bits of the float32 bit pattern.
+        All special values (NaN, infinity, zero, subnormals) are preserved.
+        """
+        v = float(value)
+        bits = struct.unpack('I', struct.pack('f', v))[0]
+        tf32 = bits & 0xFFFFE000
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:1078]: Valid - fixed in 67b498132. TensorFloat32.toBytes() now checks for NaN whose payload is entirely in the lower 13 masked bits (exp==0xFF, mant!=0, tf32 mantissa==0) and forces tf32 |= 0x00002000 to preserve NaN status.

File: src/rogue/interfaces/memory/Block.cpp [hunk 10]
```diff
@@ -1723,6 +1723,1107 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    std::memcpy(&result, &f, sizeof(result));
+    return result;
+}
+
+uint16_t floatToBFloat16(float value) {
+    uint32_t f;
+    std::memcpy(&f, &value, sizeof(f));
+    // BFloat16 is simply the upper 16 bits of the float32 bit pattern.
+    // All special values (NaN, infinity, zero, subnormals) are preserved.
+    return static_cast<uint16_t>(f >> 16);
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid - fixed in 67b498132. floatToBFloat16() now extracts exponent and mantissa, and when exponent==0xFF, mantissa!=0, and the truncated bf16 mantissa is 0, forces bf16 |= 0x0001 so NaN cannot be mis-encoded as Inf.

File: src/rogue/interfaces/memory/Block.cpp [hunk 11]
```diff
@@ -1723,6 +1723,1107 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    return result;
+}
+
+// TensorFloat32: 1s / 8e / 10m, bias=127, same exponent as float32
+// Stored in 4 bytes; lower 13 mantissa bits are zeroed.
+
+uint32_t floatToTensorFloat32(float value) {
+    uint32_t f;
+    std::memcpy(&f, &value, sizeof(f));
+    return f & 0xFFFFE000U;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid - fixed in 67b498132. floatToTensorFloat32() now detects when exp==0x7F800000 and mant!=0 but the masked tf32 mantissa is 0, and forces tf32 |= 0x00002000 to preserve NaN status.

File: docs/src/pyrogue_tree/core/float_types_summary.rst [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+| Float6         | E3M2           | 1    | 3        | 2        | 6          | 1 byte  | +/-28.0          | ``Float6``         |
++----------------+----------------+------+----------+----------+------------+---------+------------------+--------------------+
+| Float8         | E4M3           | 1    | 4        | 3        | 8          | 1 byte  | +/-448.0         | ``Float8``         |
++----------------+----------------+------+----------+----------+------------+---------+------------------+--------------------+
+| Float16        | IEEE 754       | 1    | 5        | 10       | 16         | 2 bytes | +/-65504.0       | ``Float``          |
++----------------+----------------+------+----------+----------+------------+---------+------------------+--------------------+
+| BFloat16       | Brain Float 16 | 1    | 8        | 7        | 16         | 2 bytes | +/-3.39e38       | ``BFloat16``       |
++----------------+----------------+------+----------+----------+------------+---------+------------------+--------------------+
+| TensorFloat32  | NVIDIA TF32    | 1    | 8        | 10       | 19         | 4 bytes | +/-3.40e38       | ``TensorFloat32``  |
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/pyrogue_tree/core/float_types_summary.rst:31]: Valid - fixed in 67b498132. The Float16 row in the Model Class column now reads 'Float16' instead of 'Float'.

File: docs/src/pyrogue_tree/core/model.rst [hunk 13]
```diff
@@ -81,7 +81,8 @@ Common built-in Models include:
 * Integer: ``UInt``, ``UIntBE``, ``UIntReversed``, ``Int``, ``IntBE``
 * Boolean: ``Bool``
 * Text: ``String``
-* Floating point: ``Float``, ``FloatBE``, ``Double``, ``DoubleBE``
+* Floating point: ``Float``, ``FloatBE``, ``Float4``, ``Float4BE``, ``Float6``, ``Float6BE``, ``Float8``, ``Float8BE``, ``BFloat16``, ``BFloat16BE``, ``TensorFloat32``, ``TensorFloat32BE``, ``Double``, ``DoubleBE``.
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/pyrogue_tree/core/model.rst]: Valid - fixed in 67b498132. Float16 and Float16BE added to the floating-point model list in pyrogue_tree/core/model.rst.

File: docs/src/api/cpp/interfaces/memory/constants.rst [hunk 14]
```diff
@@ -39,5 +39,6 @@ The following block processing types are defined:
 .. doxygenvariable:: rogue::interfaces::memory::String
 .. doxygenvariable:: rogue::interfaces::memory::Float
 .. doxygenvariable:: rogue::interfaces::memory::Double
+.. doxygenvariable:: rogue::interfaces::memory::Float16
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/api/cpp/interfaces/memory/constants.rst:42]: Valid - fixed in 67b498132. Float8, BFloat16, TensorFloat32, Float6, and Float4 doxygenvariable entries added to docs/src/api/cpp/interfaces/memory/constants.rst.

File: docs/src/api/cpp/interfaces/memory/model.rst [hunk 15]
```diff
@@ -20,6 +20,7 @@ Primary API reference pages:
 - :doc:`/api/python/pyrogue/bool`
 - :doc:`/api/python/pyrogue/string`
 - :doc:`/api/python/pyrogue/float`
+- :doc:`/api/python/pyrogue/float16`
 - :doc:`/api/python/pyrogue/double`
 - :doc:`/api/python/pyrogue/fixed`
 - :doc:`/api/python/pyrogue/ufixed`
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/api/cpp/interfaces/memory/model.rst:31]: Valid - fixed in 67b498132. Added anchor sections and Python API doc links for Float8, Float8BE, BFloat16, BFloat16BE, TensorFloat32, TensorFloat32BE, Float6, Float6BE, Float4, and Float4BE in docs/src/api/cpp/interfaces/memory/model.rst.

File: tests/core/test_bfloat16.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+    # -Inf -> 0xFF80 (little-endian: [0x80, 0xFF])
+    assert model.toBytes(float('-inf')) == bytearray([0x80, 0xFF])
+
+
+def test_bfloat16_be_endianness():
+    """BFloat16BE stores bytes in big-endian order."""
+    model = pr.BFloat16BE(16)
+    assert model.endianness == 'big'
+
+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_bfloat16.py:118]: Valid - fixed in 67b498132. test_bfloat16_be_endianness() now validates byte order with known bit-pattern assertions for 1.0 (0x3F80), +Inf (0x7F80), and -Inf (0xFF80) in big-endian encoding.

File: tests/core/test_tensorfloat32.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+    # +Inf -> 0x7F800000 (little-endian: [0x00, 0x00, 0x80, 0x7F])
+    assert model.toBytes(float('inf')) == bytearray([0x00, 0x00, 0x80, 0x7F])
+    # -Inf -> 0xFF800000 (little-endian: [0x00, 0x00, 0x80, 0xFF])
+    assert model.toBytes(float('-inf')) == bytearray([0x00, 0x00, 0x80, 0xFF])
+
+
+def test_tensorfloat32_be_endianness():
+    """TensorFloat32BE stores bytes in big-endian order."""
+    model = pr.TensorFloat32BE(32)
+    assert model.endianness == 'big'
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_tensorfloat32.py:105]: Valid - fixed in 67b498132. test_tensorfloat32_be_endianness() now validates that toBytes(1.0) produces [0x3F, 0x80, 0x00, 0x00] and fromBytes round-trips correctly.

File: tests/core/test_tensorfloat32.py [hunk 18]
```diff
@@ -0,0 +1,159 @@
+#-----------------------------------------------------------------------------
+# This file is part of the rogue software platform. It is subject to
+# the license terms in the LICENSE.txt file found in the top-level directory
+# of this distribution and at:
+#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_tensorfloat32.py]: invalid

File: src/rogue/interfaces/memory/Block.cpp [hunk 19]
```diff
@@ -1723,6 +1723,1120 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+uint8_t floatToFloat6(float value) {
+    uint32_t f;
+    std::memcpy(&f, &value, sizeof(f));
+
+    uint8_t  sign     = (f >> 26) & 0x20;  // sign bit -> bit 5 of result
+    int32_t  exponent = ((f >> 23) & 0xFF) - 127 + 3;  // rebias to E3M2 bias=3
+    uint32_t mantissa = f & 0x7FFFFF;
+
+    // NaN or Infinity input -> clamp to max finite (no NaN/Inf in E3M2)
+    if (((f >> 23) & 0xFF) == 0xFF) return sign | 0x1F;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Fixed in commit 3182943c3. `floatToFloat6()` now distinguishes NaN from Inf: NaN always clamps to positive max (sign forced to 0), while Inf preserves sign. Matches Python model behavior.

File: src/rogue/interfaces/memory/Block.cpp [hunk 20]
```diff
@@ -1723,6 +1723,1120 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+uint8_t floatToFloat4(float value) {
+    uint32_t f;
+    std::memcpy(&f, &value, sizeof(f));
+
+    uint8_t  sign     = (f >> 28) & 0x08;  // sign bit -> bit 3 of result
+    int32_t  exponent = ((f >> 23) & 0xFF) - 127 + 1;  // rebias to E2M1 bias=1
+    uint32_t mantissa = f & 0x7FFFFF;
+
+    // NaN or Infinity input -> clamp to max finite (no NaN/Inf in E2M1)
+    if (((f >> 23) & 0xFF) == 0xFF) return sign | 0x07;
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Fixed in commit 3182943c3. Same fix applied to `floatToFloat4()`: NaN clamps to positive max (sign=0), Inf preserves sign. Added remote variable NaN tests for both Float4 and Float6.

File: python/pyrogue/_Model.py [hunk 21]
```diff
@@ -781,6 +781,537 @@ class DoubleBE(Double):
...
+            Byte array to extract value from.
+
+        Returns
+        -------
+        float
+            Python value.
+
+        """
+        return struct.unpack(self.fstring, ba)[0]
+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:843]: Valid observation — documented in commit 3182943c3. Added a Notes section to `Float16.toBytes()` explaining that `struct.pack` uses IEEE 754 round-to-nearest-even while the C++ `floatToHalf` uses truncation (difference is at most 1 ULP). Aligning the C++ implementation would be a separate, targeted change.

File: docs/src/pyrogue_tree/core/model.rst [hunk 22]
```diff
@@ -113,6 +114,26 @@ Built-In Model Types
 +---------------------------------------------+-----------------------+-------------------+----------------+------------------------------------------------+
 | :ref:`interfaces_memory_model_doublebe`     | 64-bit float          | float             | 64-bits        | Big endian                                     |
 +---------------------------------------------+-----------------------+-------------------+----------------+------------------------------------------------+
+| Float8                                      | 8-bit E4M3 float      | float             | 8-bits         | NVIDIA FP8 (Hopper, Blackwell). No infinity.   |
++---------------------------------------------+-----------------------+-------------------+----------------+------------------------------------------------+
+| Float8BE                                    | 8-bit E4M3 float      | float             | 8-bits         | Big endian                                     |
++---------------------------------------------+-----------------------+-------------------+----------------+------------------------------------------------+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/pyrogue_tree/core/model.rst]: Fixed in commit 3182943c3. Added `Float16` and `Float16BE` rows to the Built-In Model Types table in `model.rst`.

File: tests/core/test_tensorfloat32.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_tensorfloat32.py:5]: Fixed in the PR description — updated `test_tensor_float32.py` to the correct filename `test_tensorfloat32.py`.

File: include/rogue/interfaces/memory/Variable.h [hunk 23]
```diff
@@ -246,6 +246,66 @@ class Variable {
...
+    /////////////////////////////////
+
+    void (rogue::interfaces::memory::Block::*setFloat4_)(const float&,
+                                                         rogue::interfaces::memory::Variable*,
+                                                         int32_t index);
+
+    float (rogue::interfaces::memory::Block::*getFloat4_)(rogue::interfaces::memory::Variable*, int32_t index);
+
     /////////////////////////////////
     // C++ filed point
```
[the remark this answers is not in the document]
Comment (ruck314) [reply include/rogue/interfaces/memory/Variable.h]: Valid. Fixed in 88ef930da.

File: docs/src/pyrogue_tree/core/float_types_summary.rst [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
++----------------+----------------+------+----------+----------+------------+---------+------------------+--------------------+
+
+.. note::
+
+   Each type also has a big-endian variant (e.g., ``Float4BE``, ``Float6BE``,
+   ``Float8BE``, ``FloatBE``, ``BFloat16BE``, ``TensorFloat32BE``) for
+   registers with big-endian byte order.
```
[the remark this answers is not in the document]
Comment (ruck314) [reply docs/src/pyrogue_tree/core/float_types_summary.rst]: Valid — fixed in 88ef930da. Added Float16BE to the big-endian variant list.

File: python/pyrogue/_Model.py [hunk 25]
```diff
@@ -781,6 +781,542 @@ class DoubleBE(Double):
...
+            Byte array representation of the value.
+
+        Notes
+        -----
+        Uses ``struct.pack`` with IEEE 754 round-to-nearest-even. The C++ Block
+        path (``floatToHalf``) uses truncation instead, so values that fall on
+        a rounding boundary may differ by 1 ULP between the two paths.
+        """
+        return bytearray(struct.pack(self.fstring, value))
+
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:826]: Not fixing — this follows the existing codebase pattern. The base Float class uses bare 'f' and FloatBE uses '!f'; Double uses 'd' and DoubleBE uses '!d'. Float16 follows the same convention: 'e' for LE, '!e' for BE. While 'e' technically uses native byte order, Rogue's supported platforms (Linux x86_64, macOS arm64) are all little-endian, and changing Float16 alone would create an inconsistency with Float/Double. If explicit endian prefixes are desired, that should be a separate refactor across all float model types.

File: src/rogue/interfaces/memory/Block.cpp [hunk 26]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setFloat16",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
+
+    // Convert float to half-precision and store as 2 bytes
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2136]: Not fixing — this follows the existing framework pattern. The base `setFloat()` (line 1704) has the identical range check with `Float.minValue()=-3.4e38` / `maxValue()=3.4e38`, meaning `+/-Inf` is also rejected for regular Float variables. This is the framework's established design: range checks enforce finite representable bounds for all float types. NaN still passes through correctly (IEEE 754 NaN comparisons return false). Changing only Float16 would create inconsistency with Float/Double; if Inf pass-through is desired, that should be a separate framework-wide refactor.

File: src/rogue/interfaces/memory/Block.cpp [hunk 27]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setFloat8",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
+
+    // Convert float to E4M3 and store as 1 byte
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2276]: Not fixing — same rationale as the setFloat16 comment. The range check follows the existing `setFloat()` pattern. For Float8, values within `[-448, 448]` pass the check; values outside that range (including Inf) are rejected, which is the safer behavior for a hardware register abstraction. The clamping logic in `floatToFloat8()` is a safety net for the bit manipulation, not a user-facing feature. Consistent with Float/Double behavior.

File: src/rogue/interfaces/memory/Block.cpp [hunk 28]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+// Set data using bfloat16
+void rim::Block::setBFloat16(const float& val, rim::Variable* var, int32_t index) {
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setBFloat16",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2414]: Not fixing — same rationale. The base `setFloat()` rejects `+/-Inf` with the identical range check pattern (`Float.maxValue()=3.4e38`, so `Inf > 3.4e38` throws). BFloat16 follows this established convention. If Inf pass-through is desired across all float model types, that would be a separate framework-wide change.

File: src/rogue/interfaces/memory/Block.cpp [hunk 29]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+// Set data using TensorFloat32
+void rim::Block::setTensorFloat32(const float& val, rim::Variable* var, int32_t index) {
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setTensorFloat32",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2554]: Not fixing — same rationale as the other range check comments. Consistent with the existing `setFloat()`/`setDouble()` pattern where `+/-Inf` is rejected. NaN passes through correctly due to IEEE 754 comparison semantics.

File: src/rogue/interfaces/memory/Block.cpp [hunk 30]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setFloat6",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
+
+    // Convert float to E3M2 and store as 1 byte
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2696]: Not fixing — same rationale. The range check follows the existing `setFloat()` pattern. Float6 values within `[-28, 28]` pass; values outside are rejected. This is the safer behavior for hardware register writes. Consistent with all other float model types.

File: src/rogue/interfaces/memory/Block.cpp [hunk 31]
```diff
@@ -1723,6 +1723,1130 @@ float rim::Block::getFloat(rim::Variable* var, int32_t index) {
...
+// Set data using float4
+void rim::Block::setFloat4(const float& val, rim::Variable* var, int32_t index) {
+    // Check range
+    if ((var->minValue_ != 0 || var->maxValue_ != 0) && (val > var->maxValue_ || val < var->minValue_))
+        throw(rogue::GeneralError::create("Block::setFloat4",
+                                          "Value range error for %s. Value=%f, Min=%f, Max=%f",
+                                          var->name_.c_str(),
+                                          val,
+                                          var->minValue_,
+                                          var->maxValue_));
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp:2834]: Not fixing — same rationale. The range check follows the existing `setFloat()` pattern. Float4 values within `[-6, 6]` pass; values outside are rejected. Consistent with all other float model types in the framework.

**Review by slacrherbst — APPROVED** [review]
